src/celeritas/grid/GenericGridInserter.hh

```diff
@@ -0,0 +1,139 @@
+//---------------------------------------------------------------------------//
+//! \file celeritas/grid/GenericGridInserter.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <utility>
+#include <vector>
+
+#include "corecel/Types.hh"
+#include "corecel/cont/Span.hh"
+#include "corecel/data/Collection.hh"
+#include "corecel/data/CollectionBuilder.hh"
+#include "celeritas/Types.hh"
+#include "celeritas/io/ImportPhysicsVector.hh"
+
+#include "GenericGridBuilder.hh"
+#include "GenericGridData.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct a generic grid using mutable host data and add it to
+ * the specified grid collection.
+ *
+ * \code
+    GenericGridInserter insert(&data->reals, &data->generic_grids);
+    std::vector<GenericGridIndex> grid_ids;
+    for (material : range(MaterialId{mats->size()}))
+        grid_ids.push_back(insert(physics_vector[material.get()]));
+ */
+template<typename Index>
+class GenericGridInserter
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using SpanConstFlt = Span<float const>;
+    using SpanConstDbl = Span<double const>;
+    using RealCollection
+        = Collection<real_type, Ownership::value, MemSpace::host>;
+    using GenericGridCollection
+        = Collection<GenericGridData, Ownership::value, MemSpace::host, Index>;
+    //!@}
+
+  public:
+    //! Construct with a reference to mutable host data
+    GenericGridInserter(RealCollection* real_data, GenericGridCollection* grid);
+
+    //! Add a grid of generic data with linear interpolation
+    Index operator()(SpanConstFlt grid, SpanConstFlt values);
+
+    //! Add a grid of generic data with linear interpolation
+    Index operator()(SpanConstDbl grid, SpanConstDbl values);
+
+    //! Add an imported physics vector as a grid
+    Index operator()(ImportPhysicsVector const& vec);
+
+    //! Add an empty grid (no data present)
+    Index operator()(void);
+
+  private:
+    GenericGridBuilder grid_builder_;
+    CollectionBuilder<GenericGridData, MemSpace::host, Index> grids_;
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with a reference to mutable host data.
+ */
+template<typename Index>
+GenericGridInserter<Index>::GenericGridInserter(RealCollection* real_data,
+                                                GenericGridCollection* grid)
+    : grid_builder_(real_data), grids_(grid)
+{
+    CELER_EXPECT(real_data && grid);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Add an imported physics vector as a generic grid to the collection.
+ * Returns the id of the inserted grid, or an empty id if the vector is
+ empty.
+ */
+template<typename Index>
+auto GenericGridInserter<Index>::operator()(ImportPhysicsVector const& vec)
+    -> Index
+{
+    if (vec.x.empty())
+        return Index{};
```

Comment:
Is this expected to happen? It seems like this *could* lead to propagating unusual behavior if e.g. a physics vector isn't initialized. If the "insert an empty" is an exceptional case that only would apply in certain builders, I think it'd be better to `CELER_EXPECT(!vec.x.empty());`, and then have the caller explicitly do `if (!vec.x.empty()) { insert(vec); } else { insert(); }`.

Reply:
For some reason I thought I was mimicing a check I saw somewhere else, but it appears I was just imagining it. Changing to use `CELER_EXPECT`.